gapir/cc/memory_allocator.cpp

```diff
@@ -0,0 +1,506 @@
+      assert(inverseRelocationIter != inverseRelocationMap_.end());
+
+      auto uid = inverseRelocationIter->second;
+
+      auto relocationMapIter = relocationMap_.find(uid);
+      assert(relocationMapIter != relocationMap_.end());
+
+      auto purgableIter = purgableRegionMap_.find(relocationIter->second);
+      assert(purgableIter != purgableRegionMap_.end());
+
+      relocationMapIter->second = purgableIter;
+
+      inverseRelocationMap_.erase(inverseRelocationIter);
+      inverseRelocationMap_.insert(std::make_pair(relocationIter->second, uid));
+    }
+
+    if (purgedRegionIDs.size() > 0) {
+      auto newNullRegionIter = purgableRegionMap_.find(nullptr);
+      assert(newNullRegionIter != purgableRegionMap_.end());
+
+      for (auto&& purgedUID : purgedRegionIDs) {
+        auto iter = relocationMap_.find(purgedUID);
+        assert(iter != relocationMap_.end());
+
+        iter->second = newNullRegionIter;
+      }
+    }
+
+    purgableHead_ = newPurgableHead;
+  }
+
+  return compactedSomething;
+}
+
+size_t MemoryAllocator::getClosestStaticData(
+    size_t belowOffset,
+    std::map<unsigned char*, MemoryAllocator::MemoryRegion>::const_iterator*
+        closestStaticIter) const {
+  size_t closestStaticData = 0;
+  auto nextStaticIter = staticRegionMap_.lower_bound(heap_ + belowOffset);
+  auto prevStaticIter = nextStaticIter != staticRegionMap_.begin()
+                            ? std::prev(nextStaticIter)
+                            : staticRegionMap_.end();
+
+  if (prevStaticIter != staticRegionMap_.end()) {
+    closestStaticData =
+        prevStaticIter->second.getOffset() + prevStaticIter->second.getSize();
+  }
+
+  if (closestStaticIter != nullptr) {
+    *closestStaticIter = prevStaticIter;
+  }
+
+  return closestStaticData;
+}
+
+MemoryAllocator::Handle MemoryAllocator::registerStaticAllocate(
+    const MemoryRegion& newRegion) {
+  auto address = heap_ + newRegion.getOffset();
+  auto uid = ++idGenerator_;
+
+  auto allocIter = staticRegionMap_.insert(std::make_pair(address, newRegion));
+  inverseRelocationMap_.insert(std::make_pair(address, uid));
+  auto relocationIter =
+      relocationMap_.insert(std::make_pair(uid, allocIter.first));
+
+  return MemoryAllocator::Handle(relocationIter.first);
+}
+
+MemoryAllocator::Handle MemoryAllocator::registerPurgableAllocate(
+    const MemoryRegion& newRegion) {
+  auto address = heap_ + newRegion.getOffset();
+  auto uid = ++idGenerator_;
+
+  auto allocIter =
+      purgableRegionMap_.insert(std::make_pair(address, newRegion));
+  inverseRelocationMap_.insert(std::make_pair(address, uid));
+  auto relocationIter =
+      relocationMap_.insert(std::make_pair(uid, allocIter.first));
+
+  return MemoryAllocator::Handle(relocationIter.first);
+}
+
+void MemoryAllocator::registerResize(const MemoryRegion& resizedRegion) {
+  auto address = heap_ + resizedRegion.getOffset();
+
+  assert(staticRegionMap_.find(address) != staticRegionMap_.end());
```

Comment:
assert here only active in debug build, just checking: is that what you want?
Idem for many assert() in this file.

Reply:
Yeah. All side-effect-less